services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,514 @@
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameter is {@code null}.
+     */
+    private Future<TenantDto> findTenant(final JsonObject findQuery) {
+        Objects.requireNonNull(findQuery);
+
+        final Promise<JsonObject> findTenantPromise = Promise.promise();
+        mongoClient.findOne(config.getCollectionName(), findQuery, new JsonObject(), findTenantPromise);
+        return findTenantPromise.future()
+                .compose(tenantJsonResult -> Optional.ofNullable(tenantJsonResult)
+                        .map(tenantJson -> Future.succeededFuture(tenantJson.mapTo(TenantDto.class)))
+                        .orElse(Future.failedFuture(new NullPointerException()))
+                        .recover(error -> {
+                            if (findQuery.containsKey(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID)) {
+                                return Future.failedFuture(
+                                        new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                                String.format("Tenant [%s] not found.",
+                                                        findQuery.getString(
+                                                                RegistrationConstants.FIELD_PAYLOAD_TENANT_ID))));
+                            }
+                            return Future.failedFuture(
+                                    new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND, "Tenant not found"));
+                        }));
+    }
+
+    /**
+     * {@inheritDoc}
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>201 Created</em> if the tenant has been added successfully.</li>
+     *         <li><em>409 Conflict</em> if a tenant with the given
+     *         identifier or certificate authority already exists.</li>
+     *         <li><em>500 Internal Server Error</em> if tenants cannot be accessed.</li>
+     *         </ul>
+     * @throws NullPointerException if the parameters are {@code null}.
+     */
+    @Override
+    public Future<OperationResult<Id>> createTenant(final Optional<String> tenantId, final Tenant tenantObj,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(span);
+
+        final String tenantIdOrGenerated = tenantId.orElse(DeviceRegistryUtils.getUniqueIdentifier());
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantIdOrGenerated)
+                .compose(ok -> checkConflictingCATenants(Optional.empty(),
+                        tenantObj.getTrustedCertificateAuthoritySubjectDNs(), span))
```

Comment:
Checking for conflict CAs and creating a tenant should be atomic. Please refer the same comment as that of the update tenant and the same index should also solve the problem here. In case of create tenant, a duplicate key error could due to two reasons which are tenant id duplication or conflicting CAs. IMHO the log messages could be also adapted to say that the error could be due to two reasons.

Reply:
done